Correct branch-protection ModFlagBehavior for Aarch64 on LLVM-15

MasterAwesome opened the pull request:
When building with Fat LTO and BTI enabled on aarch64, the BTI is set to `Module::Min` for alloc shim but is set to `Module::Error` for the crate. This was fine when we were using LLVM-14 but LLVM-15 changes it's behaviour to support for compiling with different `mbranch-protection` flags.

Refer:
https://github.com/rust-lang/llvm-project/commit/b0343a38a5910e980bb031e4014655d77cd0c162

fixes https://github.com/rust-lang/rust/issues/102162

MasterAwesome committed:
Correct ModFlagBehavior for Aarch64 on LLVM-15

When building with Fat LTO and BTI enabled on aarch64, the BTI is set to
`Module::Min` for alloc shim but is set to `Module::Error` for the
crate. This was fine when we were using LLVM-14 but LLVM-15 changes it's
behaviour to support for compiling with different `mbranch-protection`
flags.

Refer:
https://github.com/rust-lang/llvm-project/commit/b0343a38a5910e980bb031e4014655d77cd0c162

nikic commented:
Min is new in LLVM 15, so this will have to be dependent on version.

Either need to pass in Min/Error depending on version, or possibly adjust https://github.com/rust-lang/rust/blob/7f42e58effa3871dda6a41e250dea60cf88868ca/compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp#L752 to automatically use Error instead of Min for old versions.

MasterAwesome commented:
Ah I see, I'll do an FFI call prior to setting the modflagbehavior to get the version then.

I could add that in the cpp file but I think it might hide future hard to debug issues where the `rustc_codegen_llvm` chooses `Min` but it's translated to `Error` in there. 

What do you think?

nikic commented:
Either way is fine by me. On the Rust side, there's already a `llvm_version` variable inside that function, so you can pick based on that (`llvm_version >= (15, 0, 0)` or so).

MasterAwesome committed:
Use `Error` behavior for LLVM versions prior to 15

CI fails when building with LLVM-13. This raises unknown behavior
constant `8` from IRVerifier.

nikic commented:
@bors r+

bors commented:
:pushpin: Commit 5480ac540cdb88a1512334c5680ccf7cd76a38c8 has been approved by `nikic`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=nikic 5480ac540cdb88a1512334c5680ccf7cd76a38c8 -->
<!-- homu: {"type":"Approved","sha":"5480ac540cdb88a1512334c5680ccf7cd76a38c8","approver":"nikic","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
